In [2]:
# Data handling
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Stats
from scipy import stats


In [5]:
df = pd.read_csv('../data/MachineLearningRating_v3.txt',
    sep='|',
    encoding='utf-8',
    engine='python'
)

## Data Understanding

In [ ]:
df.head()          # preview first 5 rows
df.info()          # check column types and missing values
df.describe()      # summary statistics for numeric columns


Based on this data snippet, here's what we can understand:

##### 🔍 **Key Observations from Insurance Data:**

##### **Data Structure:**
- **52 columns** including policy IDs, vehicle details, coverage info, and financial metrics
- **Mixed data types**: numeric, categorical, and possibly missing values (NaN)

##### **Vehicle & Policy Details:**
- **Vehicle specifications**: Registration year, cylinders, cubic capacity, kilowatts, doors
- **Coverage details**: Sum insured, premiums, claims data
- **Policy identifiers**: UnderwrittenCoverID, PolicyID, PostalCode

##### **Financial Metrics:**
- **Premium calculations**: CalculatedPremiumPerTerm, TotalPremium
- **Risk assessment**: Multiple premium/claim amount columns
- **Fleet data**: NumberOfVehiclesinFleet indicates commercial policies

##### **Data Quality Issues:**
- **Missing values**: NaN present in multiple columns
- **Formatting problems**: Scientific notation mixed with regular numbers
- **Inconsistent data**: Some columns showing extreme values or placeholders

##### **Potential Analysis Areas:**
1. **Risk modeling** using vehicle specs vs claim amounts
2. **Pricing optimization** based on calculated vs actual premiums
3. **Fraud detection** from claim patterns
4. **Customer segmentation** by postal code/vehicle type

**Next steps**: Clean the data, handle missing values, convert formats, then explore relationships between vehicle characteristics and insurance claims.

In [9]:
missing = df.isnull().sum().sort_values(ascending=False)
missing_percent = (missing / len(df)) * 100
missing_table = pd.concat([missing, missing_percent], axis=1, keys=['Missing Values', '% of Total'])
print(missing_table)


                          Missing Values  % of Total
NumberOfVehiclesInFleet          1000098  100.000000
CrossBorder                       999400   99.930207
CustomValueEstimate               779642   77.956560
Rebuilt                           641901   64.183810
Converted                         641901   64.183810
WrittenOff                        641901   64.183810
NewVehicle                        153295   15.327998
Bank                              145961   14.594670
AccountType                        40232    4.022806
Gender                              9536    0.953507
MaritalStatus                       8259    0.825819
VehicleType                          552    0.055195
make                                 552    0.055195
mmcode                               552    0.055195
Model                                552    0.055195
Cylinders                            552    0.055195
bodytype                             552    0.055195
kilowatts                            552    0.

###### 🔍 **Missing Value Analysis**

###### **Critical Issues:**
- **100% missing**: `NumberOfVehiclesInFleet` → Column is **useless**
- **99.9% missing**: `CrossBorder` → Nearly useless
- **64-78% missing**: Key vehicle history fields (`Rebuilt`, `Converted`, `WrittenOff`, `CustomValueEstimate`)

###### **Data Quality:**
- **Perfect columns**: 30+ columns have **0% missing** (policy, premium, claim data)
- **Vehicle specs problematic**: 0.06% missing in make/model/engine details
- **Demographics issues**: Gender/MaritalStatus ~1% missing

###### **Action Required:**
1. **Drop**: 100% missing columns
2. **Impute/Careful handling**: High-missing columns (>60%)
3. **Minimal impact**: Low-missing columns (<1%)

**Bottom line**: Core insurance data (premiums, claims, policy info) is **complete**, but vehicle/demographic data needs cleaning.